File: ts/src/blockchain.ts
import { Block } from "./block";

export class Blockchain {
  public chain: Block[];
  private difficulty = 2;
  constructor() {
    this.chain = [this.createGenesisBlock()];
  }

  private createGenesisBlock(): Block {
    return new Block(0);
  }

  private getLastBlock() {
    return this.chain[this.chain.length - 1];
  }

  addBlock(data: any): void {
    const prevBlock = this.getLastBlock();
    const block = new Block(prevBlock.index + 1, data, prevBlock.hash);
    block.mine(this.difficulty);
    this.chain.push(block);
  }

  isChainValid(): boolean {
    for (let i = 1; i < this.chain.length; i++) {
      const currentBlock = this.chain[i];
      const previousBlock = this.chain[i - 1];
      const tester = "0".repeat(this.difficulty);

      if (!currentBlock.hash.startsWith(tester)) {
        return false;
      }

      if (currentBlock.hash !== currentBlock.calculateHash()) {
        return false;
      }

      if (currentBlock.prevHash !== previousBlock.hash) {
        return false;
      }
    }
    return true;
  }
}

File: go/blockchain/blockchain.go
package blockchain

import (
	"blockchain-impl/go/block"
)

// Blockchain is a group of blocks linked together
type Blockchain struct {
	difficulty int
	chain      []*block.Block
	length     int
}

// Len returns the length of the blockchain
func (bc *Blockchain) Len() int {
	return bc.length
}

// AddBlock creates a block with given data and adds it to the chain
func (bc *Blockchain) AddBlock(data string) {
	prevBlock := *bc.GetLastBlock()
	newBlock := block.New(
		bc.length,
		data,
		prevBlock.Hash,
	)
	newBlock.MineBlock(bc.difficulty)
	bc.chain = append(bc.chain, newBlock)
	bc.length++
}

// GetLastBlock returns the pointer to the last block in the blockchain
func (bc *Blockchain) GetLastBlock() *block.Block {
	return bc.chain[bc.length-1]
}

func addGenesisBlock(bc *Blockchain) {
	genesis := block.New(0, "", "")
	bc.chain = append(bc.chain, genesis)
	bc.length++
}

// New returns an instance of a blockchain
func New(difficulty int) *Blockchain {
	bc := Blockchain{
		difficulty: difficulty,
	}
	addGenesisBlock(&bc)
	return &bc
}

// IsValidChain checks the validity of a chain based on hashes
func IsValidChain(bc *Blockchain) bool {
	for i := 1; i < bc.length; i++ {
		currBlock := *bc.chain[i]
		prevBlock := *bc.chain[i-1]
		currHash := currBlock.CalculateHash()
		if currBlock.Hash != currHash {
			return false
		}
		if prevBlock.Hash != currBlock.PrevHash {
			return false
		}
		return true
	}
	return true
}
